File: sdk/ruby/tests/phantex_test.rb
# frozen_string_literal: true

require "minitest/autorun"
require_relative "../lib/phantex"

class TestPhantexConfig < Minitest::Test
  def test_default_config
    config = Phantex::Config.new
    assert_equal "",            config.auth_token
    assert_equal "",            config.tenant_id
    assert_equal "auto",        config.transport
    assert_equal "localhost:50051", config.gateway_addr
    assert_equal 50,            config.batch_size
    assert_equal true,          config.enabled
    assert_equal false,         config.debug
  end

  def test_config_from_env
    ENV["PHANTEX_TOKEN"]     = "test-token-123"
    ENV["PHANTEX_TENANT_ID"] = "tenant-abc"
    ENV["PHANTEX_AGENT_ID"]  = "agent-ruby-1"
    ENV["PHANTEX_DEBUG"]     = "1"

    config = Phantex::Config.from_env
    assert_equal "test-token-123", config.auth_token
    assert_equal "tenant-abc",     config.tenant_id
    assert_equal "agent-ruby-1",   config.agent_id
    assert_equal true,             config.debug
  ensure
    ENV.delete("PHANTEX_TOKEN")
    ENV.delete("PHANTEX_TENANT_ID")
    ENV.delete("PHANTEX_AGENT_ID")
    ENV.delete("PHANTEX_DEBUG")
  end
end

class TestPhantexContext < Minitest::Test
  def test_trace_id_generation
    tid = Phantex::Context.new_trace_id
    assert_equal 32, tid.length
    assert_match(/\A[0-9a-f]+\z/, tid)
  end

  def test_span_id_generation
    sid = Phantex::Context.new_span_id
    assert_equal 16, sid.length
    assert_match(/\A[0-9a-f]+\z/, sid)
  end

  def test_agent_paid_from_env
    ENV["PHANTEX_AGENT_ID"] = "ruby-agent-99"
    Thread.current[:phantex_agent_paid] = nil
    assert_equal "ruby-agent-99", Phantex::Context.agent_paid
  ensure
    ENV.delete("PHANTEX_AGENT_ID")
  end

  def test_with_span_nesting
    Phantex::Context.span_id = "outer"
    Phantex::Context.with_span(framework_name: "test") do |inner_span|
      assert_equal inner_span, Phantex::Context.span_id
      assert_equal "outer", Phantex::Context.parent_span_id
      assert_equal "test", Phantex::Context.framework
    end
    assert_equal "outer", Phantex::Context.span_id
  end
end

class TestPhantexEvents < Minitest::Test
  def test_tool_call_event_serialization
    event = Phantex::ToolCallEvent.new(
      tool_name:  "calculator",
      protocol:   "langchainrb",
      tool_input: { expression: "2+2" },
      tenant_id:  "t-1",
    )
    h = event.to_h
    assert_equal "calculator",  h[:tool_name]
    assert_equal "langchainrb", h[:protocol]
    assert_equal "t-1",         h[:tenant_id]
    assert_equal Phantex::EventType::TOOL_CALL, h[:event_type]
    assert h[:timestamp_ns] > 0
  end

  def test_tool_response_event
    event = Phantex::ToolResponseEvent.new(
      tool_name:   "search",
      protocol:    "openai_api",
      success:     false,
      duration_ns: 42_000_000,
      error_message: "timeout",
    )
    h = event.to_h
    assert_equal false, h[:success]
    assert_equal 42_000_000, h[:duration_ns]
    assert_equal "timeout", h[:error_message]
  end

  def test_hash_prompt
    hash = Phantex.hash_prompt("hello world")
    assert_equal 64, hash.length
    assert_equal hash, Phantex.hash_prompt("hello world") # deterministic
  end
end

class TestBufferTransport < Minitest::Test
  def test_send_and_drain
    transport = Phantex::BufferTransport.new(max_size: 10)
    event = Phantex::ToolCallEvent.new(tool_name: "test", protocol: "test")
    transport.send(event)

    assert_equal 1, transport.length
    drained = transport.drain
    assert_equal 1, drained.length
    assert_equal 0, transport.length
  end

  def test_max_size_eviction
    transport = Phantex::BufferTransport.new(max_size: 2)
    3.times do |i|
      transport.send(Phantex::ToolCallEvent.new(tool_name: "t#{i}", protocol: "test"))
    end
    assert_equal 2, transport.length
  end
end

class TestPhantexClient < Minitest::Test
  def test_client_start_stop
    config = Phantex::Config.new(transport: "buffer", hooks: "none")
    transport = Phantex::BufferTransport.new
    client = Phantex::Client.new(config: config, transport: transport)

    client.start
    assert client.started

    client.stop
    refute client.started
  end

  def test_send_event
    transport = Phantex::BufferTransport.new
    config = Phantex::Config.new(transport: "buffer", hooks: "none")
    client = Phantex::Client.new(config: config, transport: transport)
    client.start

    event = Phantex::ToolCallEvent.new(tool_name: "my_tool", protocol: "custom")
    client.send_event(event)
    assert_equal 1, transport.length
  end

  def test_disabled_client
    config = Phantex::Config.new(enabled: false, hooks: "none")
    client = Phantex::Client.new(config: config)
    client.start
    refute client.started
  end
end

File: sdk/go/phantex_test.go
// SPDX-License-Identifier: Apache-2.0
// Copyright 2025-2026 The Phantex Authors

package phantex

import (
	"context"
	"encoding/json"
	"os"
	"testing"
)

func TestConfigFromEnv(t *testing.T) {
	os.Setenv("PHANTEX_TOKEN", "test-token")
	os.Setenv("PHANTEX_TENANT_ID", "tenant-123")
	os.Setenv("PHANTEX_AGENT_ID", "agent-456")
	os.Setenv("PHANTEX_TRANSPORT", "buffer")
	os.Setenv("PHANTEX_BATCH_SIZE", "100")
	os.Setenv("PHANTEX_DEBUG", "1")
	defer func() {
		os.Unsetenv("PHANTEX_TOKEN")
		os.Unsetenv("PHANTEX_TENANT_ID")
		os.Unsetenv("PHANTEX_AGENT_ID")
		os.Unsetenv("PHANTEX_TRANSPORT")
		os.Unsetenv("PHANTEX_BATCH_SIZE")
		os.Unsetenv("PHANTEX_DEBUG")
	}()

	cfg := ConfigFromEnv()
	if cfg.AuthToken != "test-token" {
		t.Errorf("expected token 'test-token', got '%s'", cfg.AuthToken)
	}
	if cfg.TenantID != "tenant-123" {
		t.Errorf("expected tenant 'tenant-123', got '%s'", cfg.TenantID)
	}
	if cfg.AgentID != "agent-456" {
		t.Errorf("expected agent 'agent-456', got '%s'", cfg.AgentID)
	}
	if cfg.Transport != "buffer" {
		t.Errorf("expected transport 'buffer', got '%s'", cfg.Transport)
	}
	if cfg.BatchSize != 100 {
		t.Errorf("expected batch size 100, got %d", cfg.BatchSize)
	}
	if !cfg.Debug {
		t.Errorf("expected debug true")
	}
}

func TestBufferTransport(t *testing.T) {
	bt := NewBufferTransport(10)

	evt := NewToolCallEvent()
	evt.ToolName = "test-tool"
	evt.Framework = "test"

	if err := bt.Send(evt); err != nil {
		t.Fatalf("send failed: %v", err)
	}

	if bt.Len() != 1 {
		t.Fatalf("expected 1 event, got %d", bt.Len())
	}

	events := bt.Drain()
	if len(events) != 1 {
		t.Fatalf("expected 1 drained event, got %d", len(events))
	}

	if bt.Len() != 0 {
		t.Fatalf("expected 0 after drain, got %d", bt.Len())
	}

	// Verify event content
	var decoded ToolCallEvent
	if err := json.Unmarshal(events[0], &decoded); err != nil {
		t.Fatalf("unmarshal failed: %v", err)
	}
	if decoded.ToolName != "test-tool" {
		t.Errorf("expected tool 'test-tool', got '%s'", decoded.ToolName)
	}
}

func TestBufferTransportOverflow(t *testing.T) {
	bt := NewBufferTransport(3)

	for i := 0; i < 5; i++ {
		evt := NewToolCallEvent()
		evt.ToolName = "tool"
		_ = bt.Send(evt)
	}

	if bt.Len() != 3 {
		t.Fatalf("expected 3 (max size), got %d", bt.Len())
	}
}

func TestContext(t *testing.T) {
	ctx := context.Background()

	// Trace ID should be generated when absent
	tid := TraceID(ctx)
	if tid == "" {
		t.Error("expected non-empty trace ID")
	}

	// Setting values
	ctx = WithTraceID(ctx, "trace-abc")
	ctx = WithSpanID(ctx, "span-def")
	ctx = WithAgentPAID(ctx, "agent-ghi")
	ctx = WithFramework(ctx, "go-openai")

	if TraceID(ctx) != "trace-abc" {
		t.Errorf("unexpected trace ID: %s", TraceID(ctx))
	}
	if SpanID(ctx) != "span-def" {
		t.Errorf("unexpected span ID: %s", SpanID(ctx))
	}
	if AgentPAID(ctx) != "agent-ghi" {
		t.Errorf("unexpected agent PAID: %s", AgentPAID(ctx))
	}
	if Framework(ctx) != "go-openai" {
		t.Errorf("unexpected framework: %s", Framework(ctx))
	}
}

func TestSpanContext(t *testing.T) {
	ctx := context.Background()
	ctx = WithTraceID(ctx, "t1")
	ctx = WithSpanID(ctx, "s1")
	ctx = WithAgentPAID(ctx, "a1")
	ctx = WithFramework(ctx, "fw1")

	sc := CurrentSpanContext(ctx)
	if sc.TraceID != "t1" || sc.SpanID != "s1" || sc.AgentPAID != "a1" || sc.Framework != "fw1" {
		t.Errorf("unexpected span context: %+v", sc)
	}
}

func TestEvents(t *testing.T) {
	call := NewToolCallEvent()
	if call.EventType != EventTypeToolCall {
		t.Errorf("expected event type %d, got %d", EventTypeToolCall, call.EventType)
	}
	if call.EventID == "" {
		t.Error("expected non-empty event ID")
	}
	if call.TimestampNs == 0 {
		t.Error("expected non-zero timestamp")
	}

	resp := NewToolResponseEvent()
	if resp.EventType != EventTypeToolResponse {
		t.Errorf("expected event type %d, got %d", EventTypeToolResponse, resp.EventType)
	}
	if !resp.Success {
		t.Error("expected success=true by default")
	}
}

func TestHashPrompt(t *testing.T) {
	h := HashPrompt("hello world")
	if len(h) != 64 {
		t.Errorf("expected 64-char hex hash, got %d chars", len(h))
	}
	// Same input should produce the same hash
	if HashPrompt("hello world") != h {
		t.Error("hash not deterministic")
	}
}

func TestClientStartStop(t *testing.T) {
	cfg := DefaultConfig()
	cfg.Transport = "buffer"
	cfg.Hooks = "none"

	client := NewClient(cfg)
	if client.Started() {
		t.Error("client should not be started yet")
	}

	if err := client.Start(); err != nil {
		t.Fatalf("start failed: %v", err)
	}
	if !client.Started() {
		t.Error("client should be started")
	}

	if err := client.Stop(); err != nil {
		t.Fatalf("stop failed: %v", err)
	}
	if client.Started() {
		t.Error("client should be stopped")
	}
}

func TestClientDisabled(t *testing.T) {
	cfg := DefaultConfig()
	cfg.Enabled = false

	client := NewClient(cfg)
	_ = client.Start()
	if client.Started() {
		t.Error("disabled client should not start")
	}
}

func TestSafeSerialize(t *testing.T) {
	result := SafeSerialize(map[string]string{"key": "value"}, 100)
	if result == "<unserializable>" {
		t.Error("expected valid JSON")
	}

	// Test truncation
	long := SafeSerialize(map[string]string{"key": "very long value that should be truncated"}, 20)
	if len(long) > 20 {
		t.Errorf("expected truncation to 20, got %d", len(long))
	}
}
